rbc/nrt.py

```diff
@@ -0,0 +1,552 @@
+import os
+from numba.core import cgutils
+from numba.cpython.hashing import _hashsecret as hashsecret
+from llvmlite import ir
+from contextlib import contextmanager
+
+void = ir.VoidType()
+i8 = ir.IntType(8)
+i8p = i8.as_pointer()
+i8pp = i8p.as_pointer()
+i32 = ir.IntType(32)
+i64 = ir.IntType(64)
+
+# { %"struct.std::atomic", void (i8*, i64, i8*)*, i8*, i8*, i64, i8* }
+# %"struct.std::atomic" = type { %"struct.std::__atomic_base" }
+# %"struct.std::__atomic_base" = type { i64 }
+
+REFCT_IDX = 0
+DTOR_IDX = 1
+DTOR_INFO_IDX = 2
+DATA_IDX = 3
+SIZE_IDX = 4
+EXTERNAL_ALLOCATOR_IDX = 5
+
+# struct MemInfo {
+#     std::atomic_size_t     refct;
+#     NRT_dtor_function dtor;
+#     void              *dtor_info;
+#     void              *data;
+#     size_t            size;    /* only used for NRT allocated memory */
+#     void* external_allocator;
+# };
+
+# atomic_i64_t is the llvm representation of std::atomic<int64>, according to godbolt.com
+atomic_i64_t = ir.global_context.get_identified_type('atomic_i64')
+atomic_i64_t.set_body(i64)
+
+# MemInfo struct
+MemInfo_t = ir.global_context.get_identified_type('Struct.MemInfo')
+# First argument ought to be "atomic_i64_t" instead of i64
+MemInfo_t.set_body(i64, i8p, i8p, i8p, i64, i8p)
+MemInfo_ptr_t = MemInfo_t.as_pointer()
+
+sizeof_MemInfo_t = i64(48)
+
+
+# Function Names
+NRT_MemInfo_init = 'NRT_MemInfo_init'
+
+NRT_MemInfo_alloc = 'NRT_MemInfo_alloc'
+NRT_MemInfo_alloc_safe = 'NRT_MemInfo_alloc_safe'
+
+NRT_MemInfo_alloc_dtor = 'NRT_MemInfo_alloc_dtor'
+NRT_MemInfo_alloc_dtor_safe = 'NRT_MemInfo_alloc_dtor_safe'
+
+NRT_Allocate = 'NRT_Allocate'
+NRT_Allocate_External = 'NRT_Allocate_External'
+NRT_MemInfo_call_dtor = 'NRT_MemInfo_call_dtor'
+
+NRT_incref = 'NRT_incref'
+NRT_decref = 'NRT_decref'
+
+NRT_MemInfo_data_fast = 'NRT_MemInfo_data_fast'
+
+NRT_MemInfo_new = 'NRT_MemInfo_new'
+NRT_MemInfo_new_varsize_dtor = 'NRT_MemInfo_new_varsize_dtor'
+NRT_MemInfo_varsize_alloc = 'NRT_MemInfo_varsize_alloc'
+NRT_MemInfo_varsize_realloc = 'NRT_MemInfo_varsize_realloc'
+NRT_MemInfo_varsize_free = 'NRT_MemInfo_varsize_free'
+NRT_MemInfo_new_varsize = 'NRT_MemInfo_new_varsize'
+NRT_Reallocate = 'NRT_Reallocate'
+NRT_dealloc = 'NRT_dealloc'
+NRT_Free = 'NRT_Free'
+NRT_MemInfo_destroy = 'NRT_MemInfo_destroy'
+
+nrt_internal_custom_dtor = 'nrt_internal_custom_dtor'
+nrt_allocate_meminfo_and_data = 'nrt_allocate_meminfo_and_data'
+allocate_varlen_buffer = 'allocate_varlen_buffer'
+malloc = 'malloc'
+free = 'free'
+realloc = 'realloc'
+
+nrt_debug_incr = 'nrt_debug_incr'
+nrt_debug_decr = 'nrt_debug_decr'
+nrt_debug_spaces = 'nrt_debug_spaces'
+nrt_global_var = '__nrt_global_var'
+
+function_types = {
+    NRT_MemInfo_init: (ir.FunctionType(void, [MemInfo_ptr_t, i8p, i64, i8p, i8p, i8p]),
+                       ('mi', 'data', 'size', 'dtor', 'dtor_info', 'external_allocator')),
```

Comment:
Some of the `i64` values here are actually `size_t`. Replace it by the correct definition.